pkg/asset/agent/manifests/nmstateconfig.go

```diff
@@ -336,3 +339,33 @@ func buildMacInterfaceMap(nmStateConfig aiv1beta1.NMStateConfig) models.MacInter
 	}
 	return macInterfaceMap
 }
+
+func checkHostCount(installConfig *types.InstallConfig, agentConfig *agentconfig.AgentConfig) {
+	numRequiredMasters, numRequiredWorkers := agent.GetReplicaCount(installConfig)
+
+	// Check that the number of replicas matches the configured hosts
+	numMasters := int64(0)
+	numWorkers := int64(0)
+	for _, host := range agentConfig.Config.Hosts {
+		switch host.Role {
+		case "master":
+			numMasters++
+		case "worker":
+			numWorkers++
+		case "":
+			// If not defined, the roles will be matched to replicas
+			if numMasters < numRequiredMasters {
+				numMasters++
+			} else if numWorkers < numRequiredWorkers {
+				numWorkers++
+			}
+		}
+	}
+
+	if numMasters != numRequiredMasters {
+		logrus.Warnf("The number of hosts configured as masters (%d) does not match the master replicas (%d)", numMasters, numRequiredMasters)
+	}
+	if numWorkers != numRequiredWorkers {
```

Comment:
Given that defining hosts is completely optional, I don't think we should have a warning in the case where `numWorkers == 0` (and likewise for `numMasters == 0` above).
It would be really unusual to define host-specific data for only _some_ hosts in a particular role, so I think in those cases it makes sense to have a warning.

Reply:
I'll add a check for numWorkers/numMasters != 0. Per the previous comment I'll add this a new master PR.